File: algo/RadixSort/radixsort_reverse.go
package main

import "fmt"
func radixsort_reverse(arr []int) (result []int) {
	maxNum := absInt(&arr[0])
	for _, value := range arr {
		if absInt(&value) > absInt(&maxNum) {
			maxNum = absInt(&value)
		}
	}
	maxDigitLen := 0
	for maxNum > 0{
		maxNum /= 10
		maxDigitLen++
	}
	dev := 1
	for digit := 0; digit < maxDigitLen; digit++ {
		counting := make([]int, 19)
		for _, value := range arr {
			radix := value / dev % 10 + 9
			counting[radix]++
		}

		counting[0] -= 1
		for index := 1; index < 19; index++ {
			counting[index] += counting[index - 1]
		}
		fmt.Println("counting after prefix = ", counting)

		result = make([]int, len(arr))

		//由前往後遍歷會遇到什麼問題? 就用註解中的for迴圈跑{211, 221}, 就能夠看到問題
		// for _, value := range arr{
		// 	radix := value / dev % 10 + 9
		// 	result[counting[radix]] = value
		// 	counting[radix]--
		// }
		// fmt.Println("result = ", result)

		for index := len(arr) - 1; index > -1; index-- {
			radix := arr[index] / dev % 10 + 9
			result[counting[radix]] = arr[index]
			counting[radix]--;
		}
		fmt.Println("result = ", result)
		arr = result
		dev *= 10
	}
	return
}

File: algo/RadixSort/radixSort_reverse.py
"""
底層用CountingSort_Reverse實現
"""
class radixSort:
    @classmethod
    def radixSort_reverse(cls, arr):
        if arr is None:
            return
        max_val = 0
        for val in arr:
            if val > max_val:
                max_val = val
        
        max_digit_length = 0
        while max_val > 0:
            max_val = max_val // 10
            max_digit_length += 1

        dev = 1
        for digit in range(max_digit_length):
            counting = [0] * 19

            for value in arr:
                print("value", value // dev  % 10)
                print("abs(value) ", abs(value) // dev % 10)
                radix = abs(value) // dev % 10 * (-1 if value < 0 else 1) + 9
                counting[radix] += 1

            counting[0] -= 1
            for index in range(1, len(counting)):
                counting[index] += counting[index- 1]
            print("counting after prefix = ", counting)

            result = [0] * len(arr)

            for index in range(len(arr) - 1, -1, -1):
                radix = radix = abs(value) // dev % 10 * (-1 if value < 0 else 1) + 9
                result[counting[radix]] = arr[index]
                counting[radix] -= 1

            # 由前往後遍歷會遇到什麼問題? 就用註解中的for迴圈跑{211, 221}, 就能夠看到問題
            # for _, value in enumerate(arr):
            #     radix = abs(value) // dev % 10 * (-1 if value < 0 else 1) + 9
            #     result[counting[radix]] = value
            #     counting[radix] -= 1
                
            dev *= 10
            arr[0: len(arr)] = result[0: len(result)]

if __name__ == "__main__":
    arr = [520, 211, 438, 888, 7, 111, 985, 666, 996, 233, 168]
    arr2 = [211, 221]
    radixSort.radixSort_reverse(arr2)
    print("ans = ", arr2)

"""
代碼改編自:
https://leetcode.cn/leetbook/read/sort-algorithms/raydw2/
概念來自:
https://www.youtube.com/watch?v=upyiryNShAs
"""
